[auto-clear] provide live feedback when a valid end_amount is input

christopherlam opened the pull request:
@cristiklein @mtalexander 

Further refinement: if a valid `end_amount` is being written, the dialog will actually demonstrate the splits to be cleared. Every change to gnc_amount_edit calls `gnc_account_get_autoclear_splits` which must terminate in a reasonable time and I'm now confident it will terminate <1s due to the failsafe if num(reachable_amounts) exceeds 10^6. This feature renders the display of newly-cleared splits in a new register obsolete, but I'd leave it because the user may wish to review them afterwards.

![image](https://user-images.githubusercontent.com/1975870/97794644-8c9cf980-1bf4-11eb-9b02-579097109e0c.png)
![image](https://user-images.githubusercontent.com/1975870/97794656-b0603f80-1bf4-11eb-9328-14ed286daf5e.png)

christopherlam commented:
Note the buildup of status string with many splits is currently very leaky. Not sure how to code this correctly.

christopherlam commented on gnucash/gnome/window-autoclear.c:
```diff
@@ -207,6 +240,7 @@ autoClearWindow (GtkWidget *parent, Account *account)
+    data->ok_button = GTK_WIDGET (gtk_builder_get_object (builder, "ok_button"));
     /* Add amount edit box */
     data->end_value = GNC_AMOUNT_EDIT(gnc_amount_edit_new());
     g_signal_connect(GTK_WIDGET(data->end_value), "activate",
```
@Bob-IT quick question -- why does the `end_value` "activate" signal trigger the `gnc_autoclear_window_ok_cb`? Shouldn't it connect to the `ok_button` instead? Is this a bug?

Bob-IT replied:
@christopherlam It is there to respond to hitting the enter key after inputting amount, otherwise one would enter amount and then need to press OK

Bob-IT commented:
> This feature renders the display of newly-cleared splits in a new register obsolete, but I'd leave it because the user may wish to review them afterwards.

@christopherlam  Why not add a check button that when ticked will show the register view after just like the SLR, default not ticked.

Bob-IT commented:
> Note the buildup of status string with many splits is currently very leaky. Not sure how to code this correctly.

@christopherlam This does not look right, I think you are leaking status. In the loop I think you need a temporary gchar. On entering the loop make a copy of status with g_strdup, then g_free status then use the copy in your g_strdup_printf and then g_free the temporary one all in the loop.

Bob-IT commented on gnucash/gnome/window-autoclear.c:
```diff
@@ -171,7 +171,48 @@ gnc_autoclear_window_cancel_cb (GtkWidget *widget,
+    }
+    else
+    {
+        gchar *status = "The following splits will be cleared:\n";
```
This needs to be translatable.

Bob-IT commented on gnucash/gnome/window-autoclear.c:
```diff
@@ -171,7 +171,48 @@ gnc_autoclear_window_cancel_cb (GtkWidget *widget,
+
+    if (errmsg)
+    {
+        gchar *status = g_strdup_printf ("Cannot clear: %s", errmsg);
```
This needs to be translatable.

christopherlam commented:
New autoclear algorithm to be more C-like and less lispy. Creates a amount => toclear_list, or amount => "dupe" hash table. I don't know how encode either a GList or a 'dupe' marker into the GHashTable value :

```scheme

(define (new-workitem hash reachable-amount elt old-val)
  (vector
   reachable-amount
   (if (or (hash-ref hash reachable-amount) (eq? old-val 'dup))
       'dup
       (cons elt old-val))))

(define (process-work hash workitem)
  (hash-set! hash
             (vector-ref workitem 0)
             (vector-ref workitem 1)))

(define (make-reachable-hash items)
  (define hash (make-hash-table))

  (for-each
   (lambda (item)
     (define worklist
       (list (new-workitem hash item item '())))

     (hash-for-each
      (lambda (k v)
        (set! worklist
          (cons (new-workitem hash (+ k item) item v) worklist)))
      hash)

     (for-each
      (lambda (workitem)
        (process-work hash workitem))
      worklist))
   items)

  hash)

(define hash (make-reachable-hash '(100 10 10 10)))

(hash-for-each
 (lambda (k v)
   (format #t "reachable=~a, to_clear=~a\n" k v))
 hash)
```

christopherlam commented:
Last commit - WIP - autoclear also accessible from Reconcile window. Idea is sound: picks up `toclear_value` from `recn->new_ending`. If error, should display in a GtkLabel. If no error, should *enable* checkbox for splits from debit & credit columns and *disable* other splits. But the latter requires traversing Debit and Credit GtkTreeView items to find the splits. Would need assistance how to navigate these TreeViews.

christopherlam commented:
> 
> 
> > This feature renders the display of newly-cleared splits in a new register obsolete, but I'd leave it because the user may wish to review them afterwards.
> 
> @christopherlam Why not add a check button that when ticked will show the register view after just like the SLR, default not ticked. I would also test to see what it looks like with a long description.

Agree best ellipsize description manually e.g. at 100 chars and sanitize string

christopherlam commented:
Ellipsizing long description. Arbitrarily maximum is 50 utf8 codepoints.

![image](https://user-images.githubusercontent.com/1975870/97874413-dc70e300-1d10-11eb-86fd-6ce6f1c85351.png)

christopherlam commented:
Merged live-feedback into maint. Removed hook into Reconcile window: too difficult to scan debit&credit lists to toggle checkboxes. For reference by a future hacker; this is a starting point -- but it'd need to prescan all uncleared splits, reset their reconcile to `NREC`, set `toclear_list` splits to `CREC`, then regenerate the reconcile dialog.

```patch
modified   gnucash/gnome/window-reconcile.c
@@ -1207,6 +1207,31 @@ gnc_ui_reconcile_window_balance_cb(GtkButton *button, gpointer data)
 }
 
 
+static void
+gnc_ui_autoclear_cb(GtkButton *button, gpointer data)
+{
+    RecnWindow *recnData = data;
+    Account *account;
+    GList *splits;
+    gchar *errmsg = NULL;
+
+    account = recn_get_account (data);
+    splits = gnc_account_get_autoclear_splits (account, recnData->new_ending, &errmsg);
+
+    if (errmsg)
+    {
+        gnc_info_dialog (NULL, "%s", errmsg);
+        g_free (errmsg);
+    }
+    else
+    {
+        for (GList *n = splits; n; n = n->next)
+            xaccSplitSetReconcile (n->data, CREC);
+        g_list_free (splits);
+        recnRefresh (recnData);
+    }
+}
+
 static void
 gnc_ui_reconcile_window_rec_cb(GtkButton *button, gpointer data)
 {
@@ -2331,6 +2356,11 @@ static GtkActionEntry recnWindow_actions [] =
         "including statement date and ending balance."),
         G_CALLBACK (gnc_ui_reconcile_window_change_cb)
     },
+    {
+        "RecnAutoclearAction", "emblem-default", N_("_Auto-clear"), "<primary>a",
+        N_("Auto-clear splits"),
+        G_CALLBACK(gnc_ui_autoclear_cb)
+    },
     {
         "RecnFinishAction", "system-run", N_("_Finish"), "<primary>w",
         N_("Finish the reconciliation of this account"),
modified   gnucash/ui/gnc-reconcile-window-ui.xml
@@ -47,6 +47,7 @@
     <separator name="ToolbarSep2"/>
     <toolitem name="AccountOpenAccount" action="AccountOpenAccountAction"/>
     <separator name="ToolbarSep3"/>
+    <toolitem name="RecnAutoclear" action="RecnAutoclearAction"/>
     <toolitem name="RecnFinish" action="RecnFinishAction"/>
     <toolitem name="RecnPostpone" action="RecnPostponeAction"/>
     <toolitem name="RecnCancel" action="RecnCancelAction"/>
```

christopherlam commented:
Note this live status feedback introduces unacceptable delay, reverted with 6a6396edc
Will try again once algorithm is modified to be https://github.com/Gnucash/gnucash/pull/813#issuecomment-720077364 thus `reachable_hash` is cached and can be quickly looked up. This C is beyond my ability for now.
